### PR TITLE
Batchnorm forward and backward optimizations

### PR DESCRIPTION
**Pull request by muralinr**
Batchnorm forward and backward optimizations

Performance improvements are available in the attached xls sheet..
[BN_configs_opt_bwd.xlsx](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5257860/BN_configs_opt_bwd.xlsx)
[BN_configs_opt_fwd.xlsx](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5257861/BN_configs_opt_fwd.xlsx)

### CODE REVIEW

File: src/ocl/batchnormocl.cpp [hunk 1]
```diff
@@ -949,8 +954,8 @@ void BatchNormBackward(Handle& handle,
             else
             {
                 variant    = 0;
-                xlocalsize = 1024;
-                xgridsize  = 1024 * c;
+                xlocalsize = 256;
```
Comment (daniellowell) [inline src/ocl/batchnormocl.cpp:957]: You have introduced a hazard here -- well expanded it. I guess it was always a hazard. 
xlocalsize = 256 which is `MIO_BN_GRP0` in the BWD pass.
The statement `if((n > 64) && (in_cstride > 160))` allows for in_cstride > 256 to pass through to variant 0 if n <=64.
where the preprocessor statement:
`#define MIO_BN_SEGTMP (MIO_BN_HW * (MIO_BN_GRP0 / MIO_BN_HW))`
produces a 0 for MIO_BN_SEGTMP
then, this statement passes the zero on:
`#define MIO_BN_SEGMENT ((MIO_BN_SEGTMP > MIO_BN_NHW) ? (MIO_BN_NHW) : (MIO_BN_SEGTMP))`
to this statement:
`#define MIO_BN_NLOOP ((MIO_BN_NHW + MIO_BN_SEGMENT - 1) / MIO_BN_SEGMENT)`
which is a divide by zero.
The goal in the original marco logic is produce a ceiling calculation to map the workitems across MIO_BN_NHW.

Comment (muralinr) [reply src/ocl/batchnormocl.cpp:957]: may be we should limit (MIO_BN_GRP0 / MIO_BN_HW) to 1 if this division produces 0.  I will make this change.

Comment (muralinr) [reply src/ocl/batchnormocl.cpp:957]: After this change, all these tests are passed for in_cstride > 256 to pass through to variant 0 if n <=64
root@miopen908-1:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -n 64 -c 2048 -H 18 -W 18 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
MIOpenDriver bnormfp16 -m 1 -s 1 -n 64 -c 2048 -H 18 -W 18 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
stats: bnormb, 0, 84942848, 84942848, 0, 617.796096, 0.412480
GPU Kernel Min Time Backwards Batch Normalization Elapsed: 0.412480 ms
Backwards prop batch norm verification passed on dx.
Backwards prop batch norm verification passed on dscale.
Backwards prop batch norm verification passed on dbias.
Backwards Prop Batch Norm Verifies on CPU and GPU.
root@miopen908-1:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -n 32 -c 2048 -H 18 -W 18 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
MIOpenDriver bnormfp16 -m 1 -s 1 -n 32 -c 2048 -H 18 -W 18 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
stats: bnormb, 0, 42475520, 42475520, 0, 727.316416, 0.175201
GPU Kernel Min Time Backwards Batch Normalization Elapsed: 0.175201 ms
Backwards prop batch norm verification passed on dx.
Backwards prop batch norm verification passed on dscale.
Backwards prop batch norm verification passed on dbias.
Backwards Prop Batch Norm Verifies on CPU and GPU.
root@miopen908-1:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -n 32 -c 2048 -H 20 -W 20 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
MIOpenDriver bnormfp16 -m 1 -s 1 -n 32 -c 2048 -H 20 -W 20 -V 1 -F 0 -b 1 -i 1 -t 1 -w 0
stats: bnormb, 0, 52436992, 52436992, 0, 912.053440, 0.172480
GPU Kernel Min Time Backwards Batch Normalization Elapsed: 0.172480 ms
Backwards prop batch norm verification passed on dx.
Backwards prop batch norm verification passed on dscale.
Backwards prop batch norm verification passed on dbias.
Backwards Prop Batch Norm Verifies on CPU and GPU.

**Comment by daniellowell** [comment]
@muralinr Looks like for small H*W cases the routing change to variant 0 works out much better.
Can you check where this config lands?

MIOpenDriver   bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1   -w 0 | 1.785267 | 1.960407 | -0.17514 | -9.8%
-- | -- | -- | -- | --

**Comment by muralinr** [comment]
> @muralinr Looks like for small H*W cases the routing change to variant 0 works out much better.
> Can you check where this config lands?
> 
> MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1 -w 0	1.785267	1.960407	-0.17514	-9.8%

yes Daniel.  Variant#0 works out much better for small H*W cases.  Above config uses variant#1. Negative performance drop is due to jitter.  My changes would not impact varaint#1.  I ran it again and see it better performance of 1.73 ms/1.77 ms instead of 1.96 ms.

root@miopen908-1:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1 -w 0
GPU Kernel Min Time Forward Batch Normalization Elapsed: 1.724973 ms
GPU Kernel Avg Time Forward Batch Normalization Elapsed: 1.737368 ms, for 99 iterations.
stats: bnormf, 0, 411042816, 411042816, 0, 714.868288, 1.724973
root@miopen908-1:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n 256 -c 256 -H 56 -W 56 -V 0 -F 1 -b 0 -i 100 -t 1 -w 0
GPU Kernel Min Time Forward Batch Normalization Elapsed: 1.765760 ms
GPU Kernel Avg Time Forward Batch Normalization Elapsed: 1.774880 ms, for 99 iterations.
stats: bnormf, 0, 411042816, 411042816, 0, 698.355648, 1.765760

**Review by daniellowell — APPROVED** [review]
